File: engine/runtime/src/main/scala/org/enso/compiler/pass/desugar/SectionsToBinOp.scala
package org.enso.compiler.pass.desugar

import org.enso.compiler.context.{FreshNameSupply, InlineContext, ModuleContext}
import org.enso.compiler.core.ir.{
  CallArgument,
  DefinitionArgument,
  Expression,
  Function,
  Module,
  Name
}
import org.enso.compiler.core.CompilerError
import org.enso.compiler.core.ir.expression.{Application, Section}
import org.enso.compiler.pass.IRPass
import org.enso.compiler.pass.analyse._
import org.enso.compiler.pass.lint.UnusedBindings

/** This pass converts operator sections to applications of binary operators.
  *
  * This pass has no configuration.
  *
  * This pass requires the context to provide:
  *
  * - A [[FreshNameSupply]].
  */
case object SectionsToBinOp extends IRPass {
  override type Metadata = IRPass.Metadata.Empty
  override type Config   = IRPass.Configuration.Default

  override val precursorPasses: Seq[IRPass] = List(
    GenerateMethodBodies
  )
  override val invalidatedPasses: Seq[IRPass] = List(
    AliasAnalysis,
    CachePreferenceAnalysis,
    DataflowAnalysis,
    DemandAnalysis,
    TailCall,
    UnusedBindings
  )

  /** Performs section to binary operator conversion on an IR module.
    *
    * @param ir the Enso IR to process
    * @param moduleContext a context object that contains the information needed
    *                      to process a module
    * @return `ir`, possibly having made transformations or annotations to that
    *         IR.
    */
  override def runModule(
    ir: Module,
    moduleContext: ModuleContext
  ): Module =
    ir.mapExpressions(
      runExpression(
        _,
        new InlineContext(
          moduleContext,
          freshNameSupply = moduleContext.freshNameSupply,
          compilerConfig  = moduleContext.compilerConfig
        )
      )
    )

  /** Performs section to binary operator conversion on an IR expression.
    *
    * @param ir the Enso IR to process
    * @param inlineContext a context object that contains the information needed
    *                      for inline evaluation
    * @return `ir`, possibly having made transformations or annotations to that
    *         IR.
    */
  override def runExpression(
    ir: Expression,
    inlineContext: InlineContext
  ): Expression = {
    val freshNameSupply = inlineContext.freshNameSupply.getOrElse(
      throw new CompilerError(
        "A fresh name supply is required for sections desugaring."
      )
    )

    ir.transformExpressions { case sec: Section =>
      desugarSections(sec, freshNameSupply)
    }
  }

  /** Desugars operator sections to fully-saturated applications of operators.
    *
    * For a left sections it will generate a partially-applied function. For
    * right sections it will generate a lambda. For sides sections it is forced
    * to generate a lambda returning a partially applied function as we do not
    * currently support partial application without the this argument.
    *
    * @param section the section to desugar
    * @return the result of desugaring `section`
    */
  private def desugarSections(
    section: Section,
    freshNameSupply: FreshNameSupply
  ): Expression = {
    section match {
      case Section.Left(arg, op, loc, passData, diagnostics) =>
        val rightArgName = freshNameSupply.newName()
        val rightCallArg =
          CallArgument.Specified(None, rightArgName, None)
        val rightDefArg = DefinitionArgument.Specified(
          rightArgName.duplicate(),
          None,
          None,
          suspended = false,
          None
        )

        if (arg.value.isInstanceOf[Name.Blank]) {
          val leftArgName = freshNameSupply.newName()
          val leftCallArg =
            CallArgument.Specified(None, leftArgName, None)
          val leftDefArg = DefinitionArgument.Specified(
            leftArgName.duplicate(),
            None,
            None,
            suspended = false,
            None
          )
          val opCall = Application.Prefix(
            function             = op,
            arguments            = List(leftCallArg, rightCallArg),
            hasDefaultsSuspended = false,
            location             = None,
            passData,
            diagnostics
          )

          val rightLam = Function.Lambda(
            List(rightDefArg),
            opCall,
            None
          )

          Function.Lambda(
            List(leftDefArg),
            rightLam,
            loc
          )

        } else {
          Application.Prefix(
            function             = op,
            arguments            = List(arg),
            hasDefaultsSuspended = false,
            location             = loc,
            passData,
            diagnostics
          )
        }

      case Section.Sides(op, loc, passData, diagnostics) =>
        val leftArgName = freshNameSupply.newName()
        val leftCallArg =
          CallArgument.Specified(None, leftArgName, None)
        val leftDefArg = DefinitionArgument.Specified(
          leftArgName.duplicate(),
          None,
          None,
          suspended = false,
          None
        )

        val rightArgName = freshNameSupply.newName()
        val rightCallArg =
          CallArgument.Specified(None, rightArgName, None)
        val rightDefArg = DefinitionArgument.Specified(
          rightArgName.duplicate(),
          None,
          None,
          suspended = false,
          None
        )

        val opCall = Application.Prefix(
          function             = op,
          arguments            = List(leftCallArg, rightCallArg),
          hasDefaultsSuspended = false,
          location             = None,
          passData,
          diagnostics
        )

        val rightLambda = Function.Lambda(
          List(rightDefArg),
          opCall,
          None
        )

        Function.Lambda(
          List(leftDefArg),
          rightLambda,
          loc
        )

      /* Note [Blanks in Sections]
       * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       * While the naiive compositional translation of `(- _)` first translates
       * the section into a function applying `-` to two arguments, one of which
       * is a blank, the compositional nature of the blanks translation actually
       * works against us here.
       *
       * As the `LambdaShorthandToLambda` pass can only operate on the
       * application with the blanks, it can't know to push the blank outside
       * that application chain. To that end, we have to handle this case
       * specially here instead. What we want it to translate to is as follows:
       *
       * `(- _)` == `x -> (- x)` == `x -> y -> y - x`
       *
       * We implement this special case here.
       *
       * The same is true of left sections.
       */

      case Section.Right(op, arg, loc, passData, diagnostics) =>
        val leftArgName = freshNameSupply.newName()
        val leftCallArg =
          CallArgument.Specified(None, leftArgName, None)
        val leftDefArg =
          DefinitionArgument.Specified(
            leftArgName.duplicate(),
            None,
            None,
            suspended = false,
            None
          )

        if (arg.value.isInstanceOf[Name.Blank]) {
          // Note [Blanks in Sections]
          val rightArgName = freshNameSupply.newName()
          val rightCallArg =
            CallArgument.Specified(None, rightArgName, None)
          val rightDefArg = DefinitionArgument.Specified(
            rightArgName.duplicate(),
            None,
            None,
            suspended = false,
            None
          )

          val opCall = Application.Prefix(
            function             = op,
            arguments            = List(leftCallArg, rightCallArg),
            hasDefaultsSuspended = false,
            location             = None,
            passData,
            diagnostics
          )

          val leftLam = Function.Lambda(
            List(leftDefArg),
            opCall,
            None
          )

          Function.Lambda(
            List(rightDefArg),
            leftLam,
            loc
          )
        } else {
          val opCall = Application.Prefix(
            function             = op,
            arguments            = List(leftCallArg, arg),
            hasDefaultsSuspended = false,
            location             = None,
            passData,
            diagnostics
          )

          Function.Lambda(
            List(leftDefArg),
            opCall,
            loc
          )
        }
    }
  }
}

File: engine/runtime-compiler/src/main/java/org/enso/compiler/pass/desugar/SectionsToBinOp.java
package org.enso.compiler.pass.desugar;

import java.util.List;
import org.enso.compiler.context.InlineContext;
import org.enso.compiler.context.ModuleContext;
import org.enso.compiler.core.ir.CallArgument;
import org.enso.compiler.core.ir.DefinitionArgument;
import org.enso.compiler.core.ir.Expression;
import org.enso.compiler.core.ir.Function;
import org.enso.compiler.core.ir.MetadataStorage;
import org.enso.compiler.core.ir.Name;
import org.enso.compiler.core.ir.expression.Application;
import org.enso.compiler.core.ir.expression.Section;
import org.enso.compiler.pass.IRProcessingPass;
import org.enso.compiler.pass.MiniIRPass;
import org.enso.compiler.pass.MiniPassFactory;
import org.enso.compiler.pass.analyse.AliasAnalysis$;
import org.enso.compiler.pass.analyse.CachePreferenceAnalysis$;
import org.enso.compiler.pass.analyse.DataflowAnalysis$;
import org.enso.compiler.pass.analyse.DemandAnalysis$;
import org.enso.compiler.pass.analyse.TailCall;
import org.enso.compiler.pass.lint.UnusedBindings$;
import org.enso.persist.Persistance;
import scala.Option;
import scala.collection.immutable.Seq;
import scala.jdk.javaapi.CollectionConverters;

public final class SectionsToBinOp implements MiniPassFactory {

  public static final SectionsToBinOp INSTANCE = new SectionsToBinOp();

  private SectionsToBinOp() {}

  @Override
  public Seq<IRProcessingPass> precursorPasses() {
    List<IRProcessingPass> passes = List.of(GenerateMethodBodies$.MODULE$);
    return CollectionConverters.asScala(passes).toList();
  }

  @Override
  public Seq<IRProcessingPass> invalidatedPasses() {
    List<IRProcessingPass> passes =
        List.of(
            AliasAnalysis$.MODULE$,
            CachePreferenceAnalysis$.MODULE$,
            DataflowAnalysis$.MODULE$,
            DemandAnalysis$.MODULE$,
            TailCall.INSTANCE,
            UnusedBindings$.MODULE$);
    return CollectionConverters.asScala(passes).toList();
  }

  @Override
  public MiniIRPass createForModuleCompilation(ModuleContext moduleContext) {
    var ctx = InlineContext.fromModuleContext(moduleContext);
    return new Mini(ctx);
  }

  @Override
  public MiniIRPass createForInlineCompilation(InlineContext inlineContext) {
    return new Mini(inlineContext);
  }

  private static final class Mini extends MiniIRPass {

    private final InlineContext ctx;

    private Mini(InlineContext ctx) {
      this.ctx = ctx;
    }

    public Expression transformExpression(Expression ir) {
      var freshNameSupply = ctx.freshNameSupply().get();
      return switch (ir) {
        case Section.Left sectionLeft -> {
          var arg = sectionLeft.arg();
          var op = sectionLeft.operator();
          var loc = sectionLeft.location().isDefined() ? sectionLeft.location().get() : null;
          var passData = sectionLeft.passData();
          var rightArgName = freshNameSupply.newName(false, Option.empty());
          var rightCallArg =
              CallArgument.Specified.builder()
                  .name(Option.empty())
                  .value(rightArgName)
                  .isSynthetic(true)
                  .build();
          var rightDefArg =
              DefinitionArgument.Specified.builder()
                  .name(rightArgName.duplicate(true, true, true, false))
                  .suspended(false)
                  .build();

          if (arg.value() instanceof Name.Blank) {
            var leftArgName = freshNameSupply.newName(false, Option.empty());
            var leftCallArg =
                CallArgument.Specified.builder()
                    .name(Option.empty())
                    .value(leftArgName)
                    .isSynthetic(true)
                    .build();
            var leftDefArg =
                DefinitionArgument.Specified.builder()
                    .name(leftArgName.duplicate(true, true, true, false))
                    .suspended(false)
                    .build();
            var opCall =
                Application.Prefix.builder()
                    .function(op)
                    .arguments(cons(leftCallArg, cons(rightCallArg, nil())))
                    .hasDefaultsSuspended(false)
                    .passData(passData)
                    .diagnostics(sectionLeft.diagnostics())
                    .build();

            var rightLam =
                Function.Lambda.builder()
                    .arguments(cons(rightDefArg, nil()))
                    .bodyReference(Persistance.Reference.of(opCall))
                    .canBeTCO(true)
                    .build();

            yield Function.Lambda.builder()
                .arguments(cons(leftDefArg, nil()))
                .bodyReference(Persistance.Reference.of(rightLam))
                .canBeTCO(true)
                .location(loc)
                .build();
          } else {
            yield Application.Prefix.builder()
                .function(op)
                .arguments(cons(arg, nil()))
                .hasDefaultsSuspended(false)
                .location(loc)
                .passData(passData)
                .diagnostics(sectionLeft.diagnostics())
                .build();
          }
        }

        case Section.Sides sectionSides -> {
          var op = sectionSides.operator();
          var loc = sectionSides.location().isDefined() ? sectionSides.location().get() : null;
          var passData = sectionSides.passData();
          var leftArgName = freshNameSupply.newName(false, Option.empty());
          var leftCallArg =
              CallArgument.Specified.builder()
                  .name(Option.empty())
                  .value(leftArgName)
                  .isSynthetic(true)
                  .build();
          var leftDefArg =
              DefinitionArgument.Specified.builder()
                  .name(leftArgName.duplicate(true, true, true, false))
                  .suspended(false)
                  .build();

          var rightArgName = freshNameSupply.newName(false, Option.empty());
          var rightCallArg =
              CallArgument.Specified.builder()
                  .name(Option.empty())
                  .value(rightArgName)
                  .isSynthetic(true)
                  .build();
          var rightDefArg =
              DefinitionArgument.Specified.builder()
                  .name(rightArgName.duplicate(true, true, true, false))
                  .suspended(false)
                  .build();

          var opCall =
              Application.Prefix.builder()
                  .function(op)
                  .arguments(cons(leftCallArg, cons(rightCallArg, nil())))
                  .hasDefaultsSuspended(false)
                  .passData(passData)
                  .diagnostics(sectionSides.diagnostics())
                  .build();

          var rightLambda =
              Function.Lambda.builder()
                  .arguments(cons(rightDefArg, nil()))
                  .bodyReference(Persistance.Reference.of(opCall))
                  .canBeTCO(true)
                  .build();

          yield Function.Lambda.builder()
              .arguments(cons(leftDefArg, nil()))
              .bodyReference(Persistance.Reference.of(rightLambda))
              .canBeTCO(true)
              .location(loc)
              .build();
        }

        /* Note [Blanks in Sections]
         * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         * While the naiive compositional translation of `(- _)` first translates
         * the section into a function applying `-` to two arguments, one of which
         * is a blank, the compositional nature of the blanks translation actually
         * works against us here.
         *
         * As the `LambdaShorthandToLambda` pass can only operate on the
         * application with the blanks, it can't know to push the blank outside
         * that application chain. To that end, we have to handle this case
         * specially here instead. What we want it to translate to is as follows:
         *
         * `(- _)` == `x -> (- x)` == `x -> y -> y - x`
         *
         * We implement this special case here.
         *
         * The same is true of left sections.
         */

        case Section.Right sectionRight -> {
          var arg = sectionRight.arg();
          var op = sectionRight.operator();
          var loc = sectionRight.location().isDefined() ? sectionRight.location().get() : null;
          var passData = sectionRight.passData();
          var leftArgName = freshNameSupply.newName(false, Option.empty());
          var leftCallArg =
              CallArgument.Specified.builder()
                  .name(Option.empty())
                  .value(leftArgName)
                  .isSynthetic(true)
                  .build();
          var leftDefArg =
              DefinitionArgument.Specified.builder()
                  .name(leftArgName.duplicate(true, true, true, false))
                  .suspended(false)
                  .build();

          if (arg.value() instanceof Name.Blank) {
            // Note [Blanks in Sections]
            var rightArgName = freshNameSupply.newName(false, Option.empty());
            var rightCallArg =
                CallArgument.Specified.builder()
                    .name(Option.empty())
                    .value(rightArgName)
                    .isSynthetic(true)
                    .build();
            var rightDefArg =
                DefinitionArgument.Specified.builder()
                    .name(rightArgName.duplicate(true, true, true, false))
                    .suspended(false)
                    .build();

            var opCall =
                Application.Prefix.builder()
                    .function(op)
                    .arguments(cons(leftCallArg, cons(rightCallArg, nil())))
                    .hasDefaultsSuspended(false)
                    .passData(passData)
                    .diagnostics(sectionRight.diagnostics())
                    .build();

            var leftLam =
                Function.Lambda.builder()
                    .arguments(cons(leftDefArg, nil()))
                    .bodyReference(Persistance.Reference.of(opCall))
                    .canBeTCO(true)
                    .build();

            yield Function.Lambda.builder()
                .arguments(cons(rightDefArg, nil()))
                .bodyReference(Persistance.Reference.of(leftLam))
                .canBeTCO(true)
                .location(loc)
                .build();
          } else {
            var opCall =
                Application.Prefix.builder()
                    .function(op)
                    .arguments(cons(leftCallArg, cons(arg, nil())))
                    .hasDefaultsSuspended(false)
                    .passData(passData)
                    .diagnostics(sectionRight.diagnostics())
                    .build();

            yield Function.Lambda.builder()
                .arguments(cons(leftDefArg, nil()))
                .bodyReference(Persistance.Reference.of(opCall))
                .canBeTCO(true)
                .location(loc)
                .build();
          }
        }
        default -> ir;
      };
    }

    private static MetadataStorage meta() {
      return new MetadataStorage();
    }

    @SuppressWarnings("unchecked")
    private static <T> scala.collection.immutable.List<T> nil() {
      return (scala.collection.immutable.List<T>) scala.collection.immutable.Nil$.MODULE$;
    }

    private static <T> scala.collection.immutable.List<T> cons(
        T head, scala.collection.immutable.List<T> tail) {
      return scala.collection.immutable.$colon$colon$.MODULE$.apply(head, tail);
    }
  }
}
